File: EPI/Heaps/Implement a stack API using a heap/swift/11.7.swift
class Stack<T> {
    let maxIndexHeap = Heap<(index: Int, value: T)> { $0.index > $1.index }

    func push(_ element: T) {
        maxIndexHeap.insert((index: maxIndexHeap.count, value: element))
    }

    func pop() -> T? {
        return maxIndexHeap.removeHighestPriority()?.value
    }
}

File: EPI/Heaps/Implement a stack API using a heap/typescript/11.7.ts
import Heap from './Heap';

interface IndexedElement<T> {
    index: number;
    element: T;
}

export default class Stack<T> {
    private heap = new Heap<IndexedElement<T>>((i, j) => j.index - i.index);

    push(element: T): void {
        this.heap.insert({
            index: this.heap.size,
            element,
        });
    }

    pop(): T | undefined {
        const indexElement = this.heap.remove();
        return indexElement && indexElement.element;
    }
}
